src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs

```diff
@@ -170,62 +170,86 @@ private static Task<Solution> GetSolutionAsync(RoslynServices roslynService, Pin
             return solutionController.GetSolutionAsync(solutionInfo.SolutionChecksum, solutionInfo.FromPrimaryBranch, cancellationToken);
         }
 
-        protected async Task<T> RunServiceAsync<T>(Func<Task<T>> callAsync, CancellationToken cancellationToken)
+        protected async Task<T> RunServiceAsync<T>(Func<CancellationToken, Task<T>> callAsync, CancellationToken cancellationToken)
         {
             AssetStorage.UpdateLastActivityTime();
 
-            try
+            // merge given cancellation token with shutdown cancellation token. it looks like if cancellation and disconnection happens
+            // almost same time, we might not get cancellation message back from stream json rpc and get disconnected
+            // https://github.com/Microsoft/vs-streamjsonrpc/issues/64
+            using (var mergedCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ShutdownCancellationToken))
             {
-                return await callAsync().ConfigureAwait(false);
-            }
-            catch (Exception ex) when (LogUnlessCanceled(ex, cancellationToken))
-            {
-                // never reach
-                return default(T);
+                try
+                {
+                    return await callAsync(mergedCancellation.Token).ConfigureAwait(false);
+                }
+                catch (Exception ex) when (LogUnlessCanceled(ex, mergedCancellation.Token))
+                {
+                    // never reach
+                    return default(T);
```

Comment:
:bulb: I believe we have `throw ExceptionHelpers.Unreachable;` for this (applies to multiple locations here and below).

Reply:
sure